src/Interpreters/HashJoin.cpp

```diff
@@ -1990,10 +2056,20 @@ void HashJoin::reuseJoinedData(const HashJoin & join)
     }
 }
 
-BlocksList HashJoin::releaseJoinedBlocks()
+BlocksList HashJoin::releaseJoinedBlocks(bool restructure)
 {
+    LOG_DEBUG(log, "Join data is being released, {} bytes and {} rows in hash table", getTotalByteCount(), getTotalRowCount());
+
     BlocksList right_blocks = std::move(data->blocks);
-    data->released = true;
+    if (!restructure)
```

Comment:
We don't need to restructure them, because we are going to insert this blocks back to another hash table. Also we will store restructured (prepared) block in buckets.